impersonate/middleware.py

```diff
@@ -1,26 +1,55 @@
+from django.utils.functional import empty, SimpleLazyObject
 from .helpers import User, check_allow_for_uri, check_allow_for_user
 
 
-class ImpersonateMiddleware(object):
+def apply_impersonate(request):
+    request.user.is_impersonate = False
+    request.__dict__['impersonator'] = None
+
+    if request.user.is_authenticated() and \
+       '_impersonate' in request.session:
+        new_user_id = request.session['_impersonate']
+        if isinstance(new_user_id, User):
+            # Edge case for issue 15
+            new_user_id = new_user_id.id
+
+        try:
+            new_user = User.objects.get(id=new_user_id)
+        except User.DoesNotExist:
+            return
+
+        if check_allow_for_user(request, new_user) and \
+           check_allow_for_uri(request.path):
+            request.__dict__['impersonator'] = request.user
+            request.user = new_user
+            request.user.is_impersonate = True
+
+
+def impersonator(request):
+    # trigger apply_impersonate
+    request.user.is_authenticated()
+
+    return request.impersonator
 
+
+class ImpersonateMiddleware(object):
     def process_request(self, request):
-        request.user.is_impersonate = False
-        request.impersonator = None
-
-        if request.user.is_authenticated() and \
-           '_impersonate' in request.session:
-            new_user_id = request.session['_impersonate']
-            if isinstance(new_user_id, User):
-                # Edge case for issue 15
-                new_user_id = new_user_id.id
-
-            try:
-                new_user = User.objects.get(pk=new_user_id)
-            except User.DoesNotExist:
-                return
-
-            if check_allow_for_user(request, new_user) and \
-               check_allow_for_uri(request.path):
-                request.impersonator = request.user
-                request.user = new_user
-                request.user.is_impersonate = True
+        # User isn't lazy, don't preserve laziness.
+        if not isinstance(request.user, SimpleLazyObject):
+            apply_impersonate(request)
+            return None
+
+        # User is already instantiated, don't preserve laziness
+        if request.user.__dict__['_wrapped'] is not empty:
+            apply_impersonate(request)
+            return None
+
+        get_user = request.user.__dict__['_setupfunc']
+
+        def wrap_user():
+            request.__dict__['user'] = get_user()
```

Comment:
I agree that it's strange that accessing request.user before setup returns a different object than after setup. Unfortunately, just setting `request.user.__dict__['_wrapped']` introduces an infinite loop, and I don't have any ideas for solving it without diving into the code.

As everything works as-is, so I'm going to leave it alone.